resources: update gapbs to use new 24.04 images

Harshil2107 opened the pull request.

Harshil2107 committed:
resources: update gapbs to use new 24.04 images

Harshil2107 committed:
resourecs: Add scripts to make arm and riscv diskimages with packer

Harshil2107 committed:
resources: Add real graphs to the disk

Harshil2107 committed:
resources: update post install to make 2nd graph work

erin-le commented on src/gapbs/README.md:
```diff
@@ -6,71 +6,112 @@ tags:
 This document provides instructions to create a GAP Benchmark Suite (GAPBS) disk image, which, along with an example script, may be used to run GAPBS within gem5 simulations. The example script uses a pre-built disk-image.
 
-A pre-built disk image, for X86, can be found, gzipped, here: <http://dist.gem5.org/dist/v22-1/images/x86/ubuntu-18-04/gapbs.img.gz>.
+A pre-built disk image, for X86, can be found, gzipped, here: <add link>
```
Add link

erin-le commented on src/gapbs/README.md:
```diff
@@ -6,71 +6,112 @@ tags:
 This document provides instructions to create a GAP Benchmark Suite (GAPBS) disk image, which, along with an example script, may be used to run GAPBS within gem5 simulations. The example script uses a pre-built disk-image.
 
-A pre-built disk image, for X86, can be found, gzipped, here: <http://dist.gem5.org/dist/v22-1/images/x86/ubuntu-18-04/gapbs.img.gz>.
+A pre-built disk image, for X86, can be found, gzipped, here: <add link>
```
Add link

Harshil2107 replied:
If everything looks good then I will upload the disks and add the links.

erin-le commented on src/gapbs/README.md:
```diff
@@ -6,71 +6,112 @@ tags:
+After this process succeeds, the disk image can be found on the `src/gapbs/disk-image-ubuntu-24-04`.
 
-GAPBS disk image can support both real and synthetic graph inputs. The current pre-built disk image contains only one graph input which includes the New York city road map (with 733K nodes) it can be found: <http://users.diag.uniroma1.it/challenge9/download.shtml>.
+This gapbs image uses the prebuilt ubuntu 24.04 image as a base image. The gapbs image also throws the same exit events as the base image.
```
Elaborate more on exit events in the base disk image and/or add link to base image documentation. 

I think you should also say "Look at the Boot Sequences section for information on what exit events will take place", then tell users where they can find configuration scripts with that exit event setup. Since the PR to update the config scripts is still pending, maybe you could point users to the "usage" tab of one of the gem5 resources with the fleshed out exit event handler setup (such as this one: https://resources.gem5.org/resources/x86-ubuntu-24.04-img/usage?database=gem5-resources&version=2.0.0).  Or even just put it in a code block somewhere in this document.

mbabaie replied:
I agree with more explanations about two things:
1. what the path to the programs' executable is in the disk image. This is probably required to setup the command in the run script. I saw in the examples Erin provided below, she has not specified that, and has  used them directly as `./cc` for this purpose.
2. number of exit events, their orders, and where they are in the main program.

Harshil2107 replied:
1. I have added the path to the executable, I will also add the links to the workloads on gem5 resources when the changes looks good, and the disks have been uploads. There will be a workload for each executable benchmark in 3 sizes and with two real graphs. Is there any other information that I should add?

2. The boot sequences section in the readme explains, where the exits will occur. will it be beneficial to mention exact files?

mbabaie replied:
1. Thanks for adding this. Looks good to me.
2. That also looks good with the stars marked to see how many exit events must be handled in the runs scripts.

erin-le commented on src/gapbs/README.md:
````diff
@@ -6,71 +6,112 @@ tags:
-```
+- The `gem5-bridge`(m5) utility is installed in `/usr/local/bin/gem5-bridge`.
+- `libm5` is installed in `/usr/local/lib/`.
+- The headers for `libm5` are installed in `/usr/local/include/gem5-bridge`.
````
Maybe you could point toward documentation for these, in case people don't know what they do/why they are important

erin-le commented on src/gapbs/README.md:
```diff
@@ -6,71 +6,112 @@ tags:
+- The `gem5-bridge`(m5) utility is installed in `/usr/local/bin/gem5-bridge`.
+- `libm5` is installed in `/usr/local/lib/`.
+- The headers for `libm5` are installed in `/usr/local/include/gem5-bridge`.
+- `gapbs` benchmark sutie with ROI annotations
```
```suggestion
- `gapbs` benchmark suite with ROI annotations
```

erin-le commented on src/gapbs/README.md:
```diff
@@ -6,71 +6,112 @@ tags:
+This gapbs image uses the prebuilt ubuntu 24.04 image as a base image. The gapbs image also throws the same exit events as the base image.
 
-To use other graphs simply copy the graph in the gapbs/ directory and add them to gapbs/gapbs.json.
+Each benchmark also has its regions of intrests annotated and they throw a `gem5-bridge workbegin` and `gem5-bridge workend` exit event.
```
```suggestion
Each benchmark also has its regions of interest annotated and they throw a `gem5-bridge workbegin` and `gem5-bridge workend` exit event.
```

erin-le commented on src/gapbs/README.md:
```diff
@@ -6,71 +6,112 @@ tags:
-* **--synthetic** refers whether to use a synthetic or a real graph. It accepts a boolean value.
-* **--size**, which refers to either the size of a synthetic graph from 1 to 16 nodes, or, a real graph. The real graph included in the pre-built disk-image is `USA-road-d.NY.gr`. Note that `--synthetic True` and `--size USA-road-d.NY.gr` cannot be combined, and, vice versa for real graphs.
+This detailed overview provides a foundational understanding of how different boot configurations affect the system's initialization and mode of operation.
+By selecting the appropriate parameters, users can customize the boot process for diverse environments, ranging from automated setups to hands-on interactive sessions.
```
It could be good to add an overview of the benchmarks available on the built disk, or point toward the GAPBS documentation

Harshil2107 replied:
I will add a section with all the benchmark workloads, when we upload the disk to the cloud and I can push the JSONs to MongoDB.

erin-le commented on src/gapbs/README.md:
```diff
@@ -6,71 +6,112 @@ tags:
+If you want to enable networking, you need to modify the disk image and move the file `/etc/netplan/00-installer-config.yaml.bak` to `/etc/netplan/00-installer-config.yaml`.
 
-To create the disk image you need to add the packer binary in the disk-image directory:
+## Building the Disk Image
```
Swap the positions of the sections "## Building the Disk Image" and "## What's on the disk?" I think this will make it more convenient for the reader, as it would put the most crucial piece of information on top.

erin-le commented on src/gapbs/build-arm.sh:
```diff
@@ -0,0 +1,10 @@
+PACKER_VERSION="1.10.0"
```
Add copyright (?)

erin-le commented on src/gapbs/build-x86.sh:
```diff
@@ -0,0 +1,18 @@
+    rm packer_${PACKER_VERSION}_linux_amd64.zip;
+fi
+
+# wget https://storage.googleapis.com/dist.gem5.org/dist/develop/images/x86/ubuntu-24-04/x86-ubuntu-24-04.gz
```
```suggestion
wget https://storage.googleapis.com/dist.gem5.org/dist/develop/images/x86/ubuntu-24-04/x86-ubuntu-24-04.gz
gunzip x86-ubuntu-24-04.gz
```

erin-le replied:
```suggestion
wget https://storage.googleapis.com/dist.gem5.org/dist/develop/images/x86/ubuntu-24-04/x86-ubuntu-24-04.gz
```

erin-le commented on src/gapbs/build-x86.sh:
```diff
@@ -0,0 +1,18 @@
+fi
+
+# wget https://storage.googleapis.com/dist.gem5.org/dist/develop/images/x86/ubuntu-24-04/x86-ubuntu-24-04.gz
+# gunzip x86-ubuntu-24-04.gz
```
```suggestion
gunzip x86-ubuntu-24-04.gz
```

erin-le commented on src/gapbs/build-arm.sh:
```diff
@@ -0,0 +1,10 @@
+    unzip packer_${PACKER_VERSION}_linux_arm64.zip;
+    rm packer_${PACKER_VERSION}_linux_arm64.zip;
+fi
+
```
```suggestion

# make the flash0.sh file
mkdir files
cd ./files

dd if=/dev/zero of=flash0.img bs=1M count=64
dd if=/usr/share/qemu-efi-aarch64/QEMU_EFI.fd of=flash0.img conv=notrunc
cd ..

wget https://storage.googleapis.com/dist.gem5.org/dist/develop/images/arm/ubuntu-24-04/arm-ubuntu-24.04-20240823.gz
gunzip arm-ubuntu-24.04-20240823.gz 

```

Harshil2107 committed:
Update src/gapbs/build-arm.sh

Co-authored-by: Erin (Jianghua) Le <jianghuale@gmail.com>

Harshil2107 committed:
resources: update path for base image

hnpl commented on src/gapbs/build-arm.sh:
```diff
@@ -0,0 +1,21 @@
+dd if=/usr/share/qemu-efi-aarch64/QEMU_EFI.fd of=flash0.img conv=notrunc
+cd ..
+
+wget https://storage.googleapis.com/dist.gem5.org/dist/develop/images/arm/ubuntu-24-04/arm-ubuntu-24.04-20240823.gz
```
This line has a problem when the .gz file already exists. Can you do one of the following?
1. If the .gz file already exists, wget will download it to .gz.1 or so. Before downloading the .gz file, make sure you remove the old download of .gz file.
2. If the .gz file already exists, checksum the file. If the checksum does not match, remove the old .gz file and re-download the .gz file.

hnpl replied:
This also applies to similar wget commands below.

erin-le commented on src/gapbs/arm-ubuntu-24-04-gapbs.pkr.hcl:
```diff
@@ -0,0 +1,76 @@
+                        ["-enable-kvm"],
+                        ["-machine", "virt"],
+                        ["-machine", "gic-version=3"],
+                        ["-vga", "virtio"],
```
```suggestion

```

erin-le commented on src/gapbs/build-riscv.sh:
```diff
@@ -1,10 +1,15 @@
     unzip packer_${PACKER_VERSION}_linux_amd64.zip;
     rm packer_${PACKER_VERSION}_linux_amd64.zip;
 fi
 
```
```suggestion
if [ ! -f riscv-ubuntu-24-04 ]; then
wget https://storage.googleapis.com/dist.gem5.org/dist/develop/images/riscv/ubuntu-24-04/riscv-ubuntu-24-04.gz;
gunzip riscv-ubuntu-24-04.gz;
fi
```

erin-le commented on src/gapbs/riscv-ubuntu-24-04-gapbs.pkr.hcl:
```diff
@@ -0,0 +1,72 @@
+  format           = "raw"
+  headless         = "true"
+  disk_image       = "true"
+  boot_command = ["<wait45>",
```
```suggestion
  boot_command = ["<wait90>",
```

erin-le commented on src/gapbs/riscv-ubuntu-24-04-gapbs.pkr.hcl:
```diff
@@ -0,0 +1,72 @@
+                  "sudo mount -o remount,rw /<enter><wait>", // explain why 
+                  "12345<enter><wait>",
+                  "sudo mv /etc/netplan/50-cloud-init.yaml.bak /etc/netplan/50-cloud-init.yaml<enter><wait>",
+                  "sudo netplan apply<enter><wait>",
```
```suggestion
                  "sudo netplan apply<enter><wait><wait>",
```

erin-le commented on src/gapbs/riscv-ubuntu-24-04-gapbs.pkr.hcl:
```diff
@@ -0,0 +1,72 @@
+                  "12345<enter><wait>",
+                  "sudo mount -o remount,rw /<enter><wait>", // explain why 
+                  "12345<enter><wait>",
+                  "sudo mv /etc/netplan/50-cloud-init.yaml.bak /etc/netplan/50-cloud-init.yaml<enter><wait>",
```
```suggestion
                  "sudo mv /etc/netplan/50-cloud-init.yaml.bak /etc/netplan/50-cloud-init.yaml<enter><wait><wait>",
```

erin-le commented on src/gapbs/riscv-ubuntu-24-04-gapbs.pkr.hcl:
```diff
@@ -0,0 +1,72 @@
+                  "gem5<enter><wait>",
+                  "12345<enter><wait>",
+                  "sudo mount -o remount,rw /<enter><wait>", // explain why 
+                  "12345<enter><wait>",
```
```suggestion
                  "12345<enter><wait><wait>",
```

erin-le commented on src/gapbs/riscv-ubuntu-24-04-gapbs.pkr.hcl:
```diff
@@ -0,0 +1,72 @@
+  boot_command = ["<wait45>",
+                  "gem5<enter><wait>",
+                  "12345<enter><wait>",
+                  "sudo mount -o remount,rw /<enter><wait>", // explain why 
```
```suggestion
                  "sudo mount -o remount,rw /<enter><wait><wait>", // explain why 
```

Harshil2107 committed:
resources: update qemu args for arm image

erin-le reviewed (commented):
I think it would also be good to add examples of how to set the workload directly into the README. It was something that I struggled with when testing your other disk image refactors. Examples for Arm, RISCV, and X86 are included below:

```python
# for RISCV
board.set_kernel_disk_workload(
    kernel=obtain_resource(
        "riscv-linux-6.6.33-kernel", resource_version="1.0.0"
    ),
    disk_image=DiskImageResource("/path/to/gem5-resources/src/gapbs/riscv-disk-image-ubuntu-24-04/riscv-ubuntu", root_partition="1"),
    bootloader=obtain_resource("riscv-bootloader-opensbi-1.3.1"), 
    readfile_contents="./cc -g 21\n",
)
```

```python
# for x86
board.set_kernel_disk_workload(
    kernel=obtain_resource(
        "x86-linux-kernel-5.4.0-105-generic", resource_version="1.0.0"
    ),
    disk_image=DiskImageResource("/path/to/gem5-resources/src/gapbs/disk-image-ubuntu-24-04/x86-ubuntu-24-04-gapbs"),
    readfile_contents="./cc -g 23\n",
    kernel_args=[
        "earlyprintk=ttyS0",
        "console=ttyS0",
        "lpj=7999923",
        "root=/dev/sda2",
        "no_systemd=true"
    ]

)
```
```python
# for arm
disk_img = DiskImageResource("/path/to/gem5-resources/src/gapbs/disk-image-arm-npb/arm-ubuntu", root_partition="2")
board.set_kernel_disk_workload(
        disk_image=disk_img, 
        bootloader=obtain_resource("arm64-bootloader-foundation"), 
        kernel=obtain_resource("arm64-linux-kernel-5.15.36"),
        readfile_contents="./cc -g 21\n",
        )

```

Harshil2107 committed:
resources: address changes about documentation

Harshil2107 committed:
resources: Update build scripts to not redownload the disk image everytime

mbabaie commented:
Related to the examples Erin has provided, I think it's good to have a full example of readfile_contents that has everything in it (whether it logs in with sudo, the full path to the binary, etc.)

Harshil2107 commented:
> Related to the examples Erin has provided, I think it's good to have a full example of readfile_contents that has everything in it (whether it logs in with sudo, the full path to the binary, etc.)

I am thinking of adding a base example scripts to run the gapbs workloads like we have right now for x86 but for all the ISAs and then point them in the readme. I think those should be good examples? is there anything else specific that I should add?

mbabaie commented:
That would be very good and helpful.
Another thing to add is to make sure we remind the users it should be run with sudo in the examples.
I don't have anything else to add at the moment :)

mbabaie commented:
This is what I meant (figured it out with trial&error) that worked perfectly:
"echo 12345 | sudo -S ./gapbs/bfs -g 22"

erin-le commented on src/gapbs/README.md:
```diff
@@ -6,71 +6,113 @@ tags:
-git clone https://gem5.googlesource.com/public/gem5
-cd gem5/util/m5
-scons build/x86/out/m5
+./build.sh          # the script downloading packer binary and building the disk image
```
```suggestion
./build-{ISA}.sh          # the script downloading packer binary and building the disk image. {ISA} is `x86`, `arm`, or `riscv`. 
```

Harshil2107 committed:
Merge branch 'gem5:stable' into add-gapbs-disk-img

erin-le reviewed (commented):
I realized that I had an old review that I never submitted, so I'm posting it now

Harshil2107 committed:
resources: Update source to use hypercalls and readme

- The change updates source to use hypercalls.
- Update readme with links to uploaded disk images.

Harshil2107 committed:
resources: Update base disk image file path for riscv

Harshil2107 committed:
resources: Update readme with example usage for gem5

erin-le commented on src/gapbs/README.md:
```diff
@@ -6,71 +6,178 @@ tags:
-* 2 Level `MESI_Two_Level` cache with 32 kB L1I and L1D size, and, 256 kB L2 size. The L1 cache(s) has associativity of 8, and, the L2 cache has associativity 16. There are 2 L2 cache banks.
-* The system has 3 GB `SingleChannelDDR4_2400` memory.
-* The script uses `x86-linux-kernel-4.19.83` and `x86-gapbs`, the disk image created from following the instructions in this `README.md`.
+**Note**: You would need to use the kernels mentions in the templates to make sure gem5 can run the workload.
```
```suggestion
**Note**: You would need to use the kernels mentioned in the templates to make sure gem5 can run the workload.
This is because these disk images are built with the [gem5 bridge driver module](https://github.com/gem5/gem5/tree/stable/util/gem5_bridge), added in [PR 1480](https://github.com/gem5/gem5/pull/1480), which makes it so that gem5 simulations don't need sudo permissions when running commands. The gem5 bridge driver module has the limitation of needing to be run with the same kernel version as what it was built with.
```

erin-le replied:
The above suggestion should also be split up so each line isn't too long.

erin-le reviewed (commented):
Thanks for the changes. I suggested adding a longer explanation about why the kernel version has to be the same; feel free to edit it for wording/ if it's inaccurate.

Harshil2107 committed:
resources: Update readme with details about gem5-bridge driver
